U/danielsf/truth/catalogs

danielsf opened the pull request:
This PR contains the code I used to generate the protoDC2 truth catalogs.  A lot of it (especially the time domain part) is not ready for public consumption (we need to have an intentional discussion about how we are going to provide truth information for time-varying sources).  I would like to merge this, however, so that I have a baseline to work against when I start developing for CosmoDC2 truth catalogs.

danielsf committed:
add script for generating example inst cats

danielsf committed:
add script to run truth writer

danielsf committed:
start trying to implement truth for sprinkler

danielsf committed:
try to get it to run

danielsf committed:
add sn_truth_params default at DB level

needed for the sprinkler to detect it

danielsf committed:
use sprinkling_switch to only output sprinkled params

danielsf committed:
write headers to truth catalogs

danielsf committed:
only output valid AGN

danielsf committed:
write un-sprinkled AGN to agn params file

danielsf committed:
factor out code to get subcatalog file name

danielsf committed:
add attempt at SQLSubCatalog mixin

danielsf committed:
rename mixin

danielsf committed:
fix SQLSubCatalogMixin so it works

danielsf committed:
make SQLSubCatalogMixin inherit from SubCatalogMixin

danielsf committed:
use SQLSubCatalogMixin

danielsf committed:
apply transformations to columns in SQL catalog

danielsf committed:
add level 7 htmid to databases

danielsf committed:
index zpoint on htmid

danielsf committed:
index on is_agn and is_sn

danielsf committed:
suppress warning about magNorm==NaN

danielsf committed:
output db name and table names

danielsf committed:
update test script

danielsf committed:
add sn_t0 to zpoint table

danielsf committed:
start writing method to make db of pointing htmid

danielsf committed:
get the trixel bounds for all of the pointings

danielsf committed:
also return a dict mapping obshistid to mjd

danielsf committed:
user specified out-dir for writing source db

danielsf committed:
update test_truth_params.py

danielsf committed:
rename TruthCatalogWriter to TruthCatalogUtils

danielsf committed:
replace assert statements with helpful exceptions

danielsf committed:
generate AGN lc (don't save them anywhere)

danielsf committed:
change prefix on sql_dir

danielsf committed:
write agn LC to hdf5 file

danielsf committed:
only simulate valid MJDs

danielsf committed:
do agn LC in chunks

danielsf committed:
need to cast ra, dec into degrees

danielsf committed:
add metadta to hdf5 file

danielsf committed:
write light curves to sqlite file, rather hdf5

danielsf committed:
only store the measured magnitude

danielsf committed:
rename method

danielsf committed:
add sne table to sqlite file

danielsf committed:
store observation metadata in table

danielsf committed:
timing statement

danielsf committed:
rename agn table to sprinkled_agn

danielsf committed:
add table for static data on objects

danielsf committed:
timing statement

danielsf committed:
build index on obs_metadata table

danielsf committed:
add quiescent magnitudes on top of dmag

danielsf committed:
add code to simulate SNe

danielsf committed:
do not attempt to INSERT if self._current_chunk is empty

danielsf committed:
use SNObject methods to calculate SN magnitudes

danielsf committed:
fix stupid bug; executemany(); not execute() in SNe

danielsf committed:
remove print statement

danielsf committed:
add ability to specify pre-existing sql_dir

danielsf committed:
timing statement

danielsf committed:
remember to apply redshift to SED

danielsf committed:
separate out sprinkled param extraction

danielsf committed:
fix timing statement

danielsf committed:
turn off dust extinction in SNe

danielsf committed:
rename tables

danielsf committed:
create indexes on LC tables

danielsf committed:
change output of write_sprinkled_params
so that we can redirect SersicTruth to the actual
database and just write the database contents then

danielsf committed:
Revert "change output of write_sprinkled_params"

This reverts commit 03edcc808a113ea964e603335fbb891bdab4cb0e.

danielsf committed:
just write Sersic magnitudes to param database

danielsf committed:
draft implementation for star truth catalog

danielsf committed:
cleaning up

danielsf committed:
back to just outputting sprinkled Sersics

danielsf committed:
add truth lc test script

danielsf committed:
configure test script to generate sprinkled db for run 1.1

danielsf committed:
add mjd back to overarching Obs_md

danielsf committed:
output parameters for all disks and bulges
we will calculate magnitudes in another script

danielsf committed:
draft script to make static galaxy db

danielsf committed:
include is_sprinkled in static galaxy catalog

danielsf committed:
fix query

danielsf committed:
remove abort

danielsf committed:
move test_truth* scripts to workspace/truth

danielsf committed:
add AGN to static galaxy truth catalog

danielsf committed:
factor out flux calculation

danielsf committed:
use correct indices for positional data

danielsf committed:
add index on galaxy_id, is_agn, is_sn to zpoint table

danielsf committed:
add missing []

danielsf committed:
get correct galaxy_id in query

danielsf committed:
convert ra, dec into degrees

danielsf committed:
add script to generate star truth catalog

danielsf committed:
query the RA, Dec columns in degrees for stars

danielsf committed:
make it possible to run scripts on pre-existing output db

danielsf committed:
fix bug (wrong name for ra, dec when building star position dict)

danielsf committed:
rename index on star table
(so that it is obvious which table it goes with)

danielsf committed:
add healpix ID to truth tables for spatial searching

danielsf committed:
fix bug

danielsf committed:
fix another bug

danielsf committed:
fix column name

danielsf committed:
expand VALUES

danielsf committed:
fix healpy import

danielsf committed:
use the correct data for healpix

danielsf committed:
fix misspelling

danielsf committed:
do not make index on initial galaxy table

danielsf committed:
rename truth scripts

danielsf committed:
write all truth to the same table

danielsf committed:
fix bug

danielsf committed:
fix bug

danielsf committed:
do not create old indexes

danielsf committed:
convert is_agn to int

danielsf committed:
use correct healpix name in index

danielsf committed:
use booleans in flag columns

danielsf committed:
change column name in index

danielsf committed:
add index on sprinkled

danielsf committed:
turns out, sqlite doesn't have boolean type; just ints

danielsf committed:
add table describing columns

danielsf committed:
fix bug

danielsf committed:
rename test_truth_lc.py to generate_truth_params.py

danielsf committed:
add comments to generate_truth_params.py

danielsf committed:
add docstring to generate_truth.py

danielsf committed:
remove test_truth_params.py

danielsf committed:
document SQLSubCatalog.py

danielsf committed:
whitespace

danielsf committed:
remove unused method

danielsf committed:
whitespace

danielsf committed:
add docstring

danielsf committed:
remove unused bp_dict

danielsf committed:
fix docstring

danielsf committed:
add comments

danielsf committed:
added docstring

danielsf committed:
edit docstring

danielsf committed:
add docstring

danielsf committed:
remove unused method

danielsf committed:
add docstring

danielsf committed:
move truth modules into main path

danielsf committed:
add docstrings to Star module

danielsf committed:
add comments

danielsf committed:
change name of kwarg input to input_db

danielsf committed:
remove commented-out code

danielsf committed:
add docstrings to galaxy module

danielsf committed:
healpy.ang2pix expects degrees when lonlat=True

danielsf committed:
fix bug

danielsf committed:
actually use chunk_size in creating star truth

danielsf committed:
demand that sprinkled params have non-null magNorm

danielsf committed:
filter galaxy magnitudes on np.isnan

danielsf committed:
fix JOIN statement in parameter query

danielsf committed:
specify the name of the sprinkled params file

danielsf committed:
specify extragalactic catalog

danielsf committed:
specify agn parameter db

danielsf committed:
add lensing parameters to sprinkled_params db

danielsf committed:
add weak lensing magnification to truth magnitudes

danielsf committed:
fix docstring

danielsf committed:
specify param_file in generate_truth.py

danielsf committed:
specify out_filein generate_truth.py

danielsf committed:
hack to deal with broken SED file names

danielsf committed:
properly broadcast magnification

danielsf committed:
add generate_truth_lc.py example

danielsf commented:
@jbkalmbach I would like to be able to pass this code on to other so that they can start thinking about how to generate a truth catalog for cosmoDC2.  Do you mind giving this PR a quick review?  Thanks.

jbkalmbach commented on workspace/truth/generate_truth_params.py:
```diff
@@ -0,0 +1,82 @@
+print('fov %e in %e sec' % (args.fov, time.time()-t_start))
+exit()
+
+### NOTHING BELOW HERE ACTUALLY HAPPENS ###
```
Is this for testing? And is it still useful?

danielsf replied:
Not exactly testing, but definitely not useful.  I am removing it.

jbkalmbach commented on python/desc/sims/GCRCatSimInterface/TruthCatalogLC.py:
```diff
@@ -0,0 +1,333 @@
+                   obshistid int, mag float)'''
+        cursor.execute(cmd)
+
+        cmd = '''CREATE TABLE obs_metadata '''
```
I don't know if this is too tough to do with sqlite in python, but maybe you could make this adaptive so that if people need extra columns from the opsim database they could add them into this table? When making light curves for Ji Won she needed `fiveSigmaDepth` and I can imagine other people might want different columns if they want to do things with the light curves.

danielsf replied:
I would like to keep the output from this as lightweight as possible.  If people want other information from the OpSimDB, they should be able to just query that database directly and use `np.in1d` or `np.searchsorted` to effectively join the two tables.  If you think that is too slow or awkward, we can revisit the question in another issue, if you don't mind.

jbkalmbach commented on python/desc/sims/GCRCatSimInterface/TruthCatalogLC.py:
```diff
@@ -0,0 +1,333 @@
+
+    with sqlite3.connect(out_file_name) as conn:
+        cursor = conn.cursor()
+        values = ((obs, mjd_dict[obs], filter_dict[obs])
```
I had to change `obs` to `int(obs)` here to get the `obsHistId` to appear correctly in the light curve db.

danielsf replied:
Ah... yes.  `sqlite` hates `np.int`.  Thanks.

jbkalmbach commented on python/desc/sims/GCRCatSimInterface/TruthCatalogLC.py:
```diff
@@ -0,0 +1,333 @@
+    agn_base_query += 'raJ2000, decJ2000, '
+    agn_base_query += 'redshift, sedFilepath, '
+    agn_base_query += 'magNorm, varParamStr '
+    agn_base_query += 'FROM zpoint WHERE is_agn=1 '
```
I think you need a `agn_base_query += 'AND is_sprinkled=1 '` here.

danielsf replied:
Either that, or I need to rename the table.  I may have originally intended this table to just contain sprinkled objects, but now I think it should contain light curves for all AGN and SNe, sprinkled or otherwise.

Do you think there is value in keeping the sprinkled light curves separate from the unsprinkled light curves, or will people be able to discern them based on position relative to the DDF?

jbkalmbach replied:
What about adding the `is_sprinkled` column and then putting them all in one table? That would make it easy to search within sqlite for just the sprinkled objects.

jbkalmbach replied:
Haha, I just noticed that's what you ended up doing. Maybe I should go through the diff first...

jbkalmbach commented on python/desc/sims/GCRCatSimInterface/TruthCatalogLC.py:
```diff
@@ -0,0 +1,333 @@
+
+            for i_chunk, agn_results in enumerate(agn_iter):
+                values = ((int(agn_results['uniqueId'][i_obj]),
+                           int(agn_results['galaxy_id'][i_obj]),
```
This needs to be `int(np.floor(agn_results['galaxy_id'][i_obj] / 10000))` for now to get the proper `galaxy_id` from a sprinkled object. And we'll have to adjust it when I finalize the number scheme for `cosmoDC2_v1.0`.

danielsf replied:
Or should we leave it as is?  If we just write the raw `galaxy_id` to the table here, that would be a way for people to identify sprinkled AGN vs. unsprinkled AGN.  I suppose they would then have to know how to disentangle the true host `galaxy_id` from the sprinkled `galaxy_id`.  Do we think people are going to want to associate the truth_catalog entries with the extragalactic catalog entries?

evevkovacs replied:
I think people might want to match truth catalog entries back to the extragalactic catalog entries, so there should be some simple way to extract the  host galaxy_id from the sprinkled galaxy_id.

jbkalmbach commented on workspace/truth/generate_truth_params.py:
```diff
@@ -0,0 +1,82 @@
+parser = argparse.ArgumentParser()
+parser.add_argument('--sql_dir', type=str, default=None)
+parser.add_argument('--sql_file', type=str, default='sprinkled_objects.sqlite')
+parser.add_argument('--fov', type=float, default=None)
```
Would be useful to add `field_ra` and `field_dec` params here.

danielsf replied:
Done

jbkalmbach reviewed (changes requested):
Skimmed the parts I haven't really interacted with and they seem fine on a quick look. Mostly focused on the AGN and sprinkled light curve code that I've used and left comments there. There's a couple changes that I think you should include (about the obsHistIds and the sprinkled galaxy_ids) and then some that are more just comments.

danielsf committed:
remove unused code

danielsf committed:
explicitly cast obsHistID into an int (rather than an np.int)

danielsf committed:
make field_ra and field_dec command line args generate_truth_params.py

danielsf commented:
If you want to come by and chat about this, feel free.  It was a very stream of consciousness design leading up to the CMU meeting.  It's possible that you are right and I should just implement your comments.

danielsf committed:
add column 'sprinkled' to object table

danielsf committed:
rename sprinkled_objects table variables_and_transients

danielsf committed:
also query is_sn=1 when getting list of valid htmid

danielsf committed:
fix typo

danielsf committed:
fix typo

danielsf committed:
alter hard-coded paths in generate_truth_lc.py

danielsf committed:
alter time statement (print hours remaining)

jbkalmbach reviewed (approved):
Looks good. Thanks for adding in the sprinkled column.

danielsf commented:
;)

Thanks for the quick review of the new commits.

You might see a few more come through in the next few hours.  I need to test that the `sprinkled` flag works correctly.  I also want to include the unsprinkled SNe before I hand this off to Eve and Yao.  Don't feel the need to review things as they come through, but if you notice something odd, let me know.

danielsf committed:
alter timing statement

danielsf committed:
remove timing statement

danielsf committed:
add sprinkled supernovae to variables_and_transients table

danielsf committed:
add module with method that will get unsprinkled SNe params

danielsf committed:
fix typo

danielsf committed:
store a full sn param dict for unsprinkled sne

danielsf committed:
edit timing statement

danielsf committed:
use 12 threads in agn simulator

danielsf committed:
apply multiprocessing to TruthCatalogLC.py SNe calcuations

danielsf committed:
guard against d_sne==0

danielsf committed:
actuall save filter as an int, rather than a string

danielsf committed:
filter light curves by what actually lands on a chip

danielsf committed:
change name of output LC database

danielsf committed:
fix typo

danielsf committed:
use angular separation to weed out things not on chip

danielsf committed:
fix typo

danielsf committed:
output of _actually_on_chip is numpy array

danielsf committed:
add print statement

danielsf committed:
typo in print statement

danielsf committed:
recall: these obs_md do not have boundLength

danielsf committed:
remove print statement

danielsf committed:
use if block to not call chipNameFromRaDecLSST unless needed

danielsf committed:
typo

danielsf committed:
remove unused lines in generate_truth_params.py

danielsf committed:
add script that actually adds unsprinkled SNe to parameter db

danielsf committed:
rename generate_truth.py to generate_truth_static.py

danielsf committed:
make add_sne.py run on arbitrary db files

danielsf committed:
move imports to top of file

danielsf committed:
move all work into __main__

danielsf committed:
generate_truth_lc.py takes command line arguments

danielsf committed:
add description of general workflow

danielsf committed:
add comments to TruthCatalogLC.py

danielsf committed:
add docstrings to TruthCatalogLC.py

danielsf committed:
do not include agn fluxes in static truth catalog

since the full AGN magnitude will be included in
the transient/variables truth catalog

danielsf committed:
remember to cast snra, sndec into radians when adding SNe to table

danielsf committed:
single light_curves table

flag whether object is AGN or SN in vairables_and_transients

danielsf committed:
actually parse args in add_sne.py

danielsf committed:
fix spelling error

yymao commented:
@danielsf I encountered an issue when trying to make the light curve reader work (cf https://github.com/LSSTDESC/gcr-catalogs/issues/203). I was assuming that all objects that are present in `variables_and_transients` would also be present in `light_curves`, but apparently that is not the case. 

Here's an example with `/global/cscratch1/sd/danielsf/Run1.2_truth/run_1.2_trial_light_curves_181004.db`:

```sql
sqlite> select uniqueId from variables_and_transients limit 1;
9262
sqlite> select obshistid from light_curves where uniqueId=9262;
sqlite> 
```

Is this expected or is this due to a bug in the truth catalog generation?

danielsf commented:
@yymao It's not unexpected (which doesn't mean it isn't also a bug)

When I populate the `variables_and_transients` table, I just write out all of the objects in our simulated population.  This neglects two details

1) To make it into the `light_curves` table, an object actually has to appear on a detector.  It is possible that there are supernovae that explode, rise, and fade before our simulated survey ever looks at them (or while it is looking away).

2) The population of simulated SNe we used for protoDC2 covers an area of sky larger than the protoDC2 area, so, even if we caught every supernova in protoDC2, there are going to be supernovae that do not appear in `light_curves` because they are outside of our survey area.

If this population of unrealized light curves is a problem for you, I can re-create the variables catalog without them (I was unaware of (2) when I wrote the code to create the truth catalog; it should be straightforward to omit the objects that run afoul of (1)).

Let me know.

yymao commented:
I can work around this in the reader now that I know it is expected. However, this means that if an user asks for the light curves of all variable objects, say within an RA range, the user may get some empty arrays for some objects. It's not very initiative but it's ok. 

How feasible is it to add a column in `variables_and_transients` table that contains the number of rows in the `light_curves` table for that object (i.e., the number of detections of that object), so that the user can select on this column?

danielsf commented:
Adding an `n_detections` column should be pretty easy.  Do you want that for Run 1.2 or should we add it to the list for Run 2.0?

yymao commented:
Doesn't matter to me actually, in terms of getting the reader to work. The reader code would be essentially the same. It's just that without this column the users won't be able to select only objects that have light curves. 

So I guess the question is whether people need this for their use cases.

danielsf commented:
If no one objects, I am going to merge this around noon PDT on October 11, just so that we have a starting point from which to start improving the truth catalog infrastructure.
